devito/types/caching.py

```diff
@@ -69,7 +69,7 @@ def __init__(self, key):
             The cache key. It must be hashable.
         """
         # Precompute hash. This uniquely depends on the cache key
-        self._cache_key_hash = hash(key)
+        self._cache_key_hash = hash(alias or key)
```

Comment:
This one is a bit odd but basically we need to make sure that all `u[t+1], u[t+2], ...` have the same cache_key_hash or in some odd case `is` will not work (== will still work but for example replace for custom coefficients breaks without this).

Reply:
I think this would be cleaner?

```
def __init__(self, key, other_keys=None):
    self._cache_key_hash = hash(key)

    awr = AugmentedWeakRef(self, self._cache_meta())
    for i in (key,) + as_tuple(other_keys):
         _SymbolCache[i] = awr
```

and should also avoid the need for calling `Cached.__init__` twice (which is conceptually wrong)

Reply:
But this is wrong as `u(x+dx)` will have a different `_cache_key_hash`

Reply:
even if you call it as:

```
Cached.__init__(newobj, newcls, other_keys=[(newcls, *indices)])
```

?